deepchem/models/torch_models/infograph.py

```diff
@@ -0,0 +1,429 @@
+        h = out.unsqueeze(0)
+
+        for i in range(3):
+            m = F.relu(self.conv(out, data.edge_index, data.edge_features))
+            out, h = self.gru(m.unsqueeze(0), h)
+            out = out.squeeze(0)
+            feat_map = out
+
+        out = self.set2set(out, data.graph_index)
+        return out, feat_map
+
+
+class InfoGraphModel(ModularTorchModel):
+    """
+    Infograph is a semi-supervised graph convolutional network for predicting molecular properties.
+    It aims to maximize the mutual information between the graph-level representation and the
+    representations of substructures of different scales. It does this by producing graph-level
+    encodings and substructure encodings, and then using a discriminator to classify if they
+    are from the same molecule or not.
+
+    To conduct training in unsupervised mode, use_unsup_loss should be True and separate_encoder
+    should be set to False. For semi-supervised training, use_unsup_loss should be True and
+    separate_encoder should be True. For supervised training, use_unsup_loss should be False.
+
+    References
+    ----------
+    F.-Y. Sun, J. Hoffmann, V. Verma, and J. Tang, “InfoGraph: Unsupervised and Semi-supervised
+    Graph-Level Representation Learning via Mutual Information Maximization.” arXiv, Jan. 17, 2020.
+    http://arxiv.org/abs/1908.01000
+
+    Parameters
+    ----------
+    num_features: int
+        Number of node features for each input
+    edge_features: int
+        Number of edge features for each input
+    dim: int
+        Dimension of the embedding
+    use_unsup_loss: bool
+        Whether to use the unsupervised loss
+    separate_encoder: bool
+        Whether to use a separate encoder for the unsupervised loss
+    measure: str
+        The divergence measure to use for the unsupervised loss. Options are 'GAN', 'JSD',
+        'KL', 'RKL', 'X2', 'DV', 'H2', or 'W1'.
+    average_loss: bool
+        Whether to average the loss over the batch
+
+    Examples
+    --------
+    >>> from deepchem.models.torch_models import InfoGraphModel
+    >>> from deepchem.feat import MolGraphConvFeaturizer
+    >>> from deepchem.data import NumpyDataset
+    >>> import torch
+    >>> smiles = ["C1CCC1", "C1=CC=CN=C1"]
+    >>> featurizer = MolGraphConvFeaturizer(use_edges=True)
+    >>> X = featurizer.featurize(smiles)
+    >>> y = torch.randint(0, 2, size=(2, 1)).float()
+    >>> w = torch.ones(size=(2, 1)).float()
+    >>> ds = NumpyDataset(X, y, w)
+    >>> num_feat = max([ds.X[i].num_node_features for i in range(len(ds))])
+    >>> edge_dim = max([ds.X[i].num_edge_features for i in range(len(ds))])
+    >>> model = InfoGraphModel(num_feat, edge_dim, 15, use_unsup_loss=True, separate_encoder=True)
+    >>> loss = model.fit(ds, nb_epoch=1)
+    """
+
+    def __init__(self,
+                 num_features,
+                 edge_features,
+                 dim,
+                 use_unsup_loss=False,
```

Comment:
I'm feeling so-so about the use of a flag to gate the unsupervised loss. Do we want to establish that as the defacto pattern for how self-supervised models turn pretraining on/off? Let's discuss offline before we make the final call

Reply:
Infograph is written with different types of training already in mind, whereas I think many models are made with only one type of training task. Modifying the loss_func is how I intend the training task to be modified in ModularTorchModel. I think in the case of infograph it makes sense, but maybe it should not be the usual pattern.